javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll

```diff
@@ -191,4 +194,37 @@ module UnsafeShellCommandConstruction {
       )
     }
   }
+
+  private import semmle.javascript.dataflow.internal.AccessPaths
+  private import semmle.javascript.dataflow.InferredTypes
+
+  /**
+   * Holds if `instance` is an instance of the access-path `ap`, and there exists a guard
+   * that ensures that `instance` is not equal to `char`.
+   */
+  private predicate blocksCharInAccess(AccessPath ap, string char, Expr instance) {
+    exists(BasicBlock bb, ConditionGuardNode guard, EqualityTest test |
+      test.getAnOperand().mayHaveStringValue(char) and
+      char = getAShellChar() and
+      guard.getTest() = test and
+      guard.dominates(bb) and
+      test.getAnOperand() = ap.getAnInstance() and
+      instance = ap.getAnInstanceIn(bb) and
+      guard.getOutcome() != test.getPolarity()
+    )
+  }
+
+  /**
+   * A sanitizer for a single char, where the char cannot be an unsafe shell char.
+   */
+  class SanitizedChar extends Sanitizer, DataFlow::ValueNode {
+    override PropAccess astNode;
+
+    SanitizedChar() {
+      exists(AccessPath ap | this.asExpr() = ap.getAnInstance() |
+        forall(string char | char = getAShellChar() | blocksCharInAccess(ap, char, astNode))
```

Comment:
Drive-by: have you seen any evidence that a shell variant of `js/incomplete-url-scheme-check` could be interesting?

Reply:
Not enough that I think we should make a query for it.   

But the [fix commit to `CVE-2020-26274`](https://github.com/sebhildebrandt/systeminformation/commit/1faadcbf68f1b1fdd5eb2054f68fc932be32ac99) was such an incomplete sanitizer.

Reply:
I have created an issue for it, but not assigned anyone.